counsel.el

```diff
@@ -3887,6 +3887,56 @@ Note: Duplicate elements of `kill-ring' are always deleted."
  '(("d" counsel-yank-pop-action-remove "delete")
    ("r" counsel-yank-pop-action-rotate "rotate")))
 
+;;** `counsel-register'
+(defvar counsel-register-actions
+  '(("\\`a buffer position" . jump-to-register)
+    ("\\`text" . insert-register)
+    ("\\`a rectangle" . insert-register)
+    ("\\`a window configuration" . jump-to-register)
+    ("\\`a frame configuration" . jump-to-register)
+    ("\\`[-+]?[0-9]+\\(?:\\.[0-9]\\)?\\'" . insert-register)
+    ("\\`the file" . jump-to-register)
+    ("\\`a keyboard macro" . jump-to-register)
+    ("\\`a file-query" . jump-to-register))
+  "Alist of (REGEXP . FUNCTION) pairs for `counsel-register'.
+Selecting a register whose description matches REGEXP specifies
+FUNCTION as the action to take on the register.")
+
+(defvar counsel-register-history nil
+  "History for `counsel-register'.")
+
+(defun counsel-register-action (register)
+  "Default action for `counsel-register'.
+
+Call a function on REGISTER.  The function is determined by
+matching the register's value description against a regexp in
+`counsel-register-actions'."
+  (setq register (string-to-char register))
```

Comment:
There is a bug here:

1. `make plain`
2. <kbd>C-x</kbd><kbd>r</kbd><kbd>SPC</kbd><kbd>C-n</kbd>
3. <kbd>M-x</kbd>`counsel-register`<kbd>RET</kbd>

   ![2019-05-13-141702_1600x900_scrot](https://user-images.githubusercontent.com/9121222/57624543-13d25980-758a-11e9-97f4-d4efa67e76fa.png)

4. <kbd>RET</kbd>

```
Debugger entered--Lisp error: (error "No action was found for register C")
  signal(error ("No action was found for register C"))
  error("No action was found for register %c" 67)
  (if (cdr action) (funcall (cdr action) register) (error "No action was found for register %c" register))
  (let* ((desc (let ((standard-output (get-buffer-create (generate-new-buffer-name " *string-output*")))) (unwind-protect (progn (let (...) (register-val-describe ... nil)) (save-current-buffer (set-buffer standard-output) (buffer-string))) (kill-buffer standard-output)))) (action (cl-assoc-if #'(lambda (re) (string-match-p re desc)) counsel-register-actions))) (if (cdr action) (funcall (cdr action) register) (error "No action was found for register %c" register)))
  counsel-register-action("C-n: buffer position: buffer *scratch*, position 1...")
  funcall(counsel-register-action "C-n: buffer position: buffer *scratch*, position 1...")
```

Here's how I propose to fix it:

```diff
diff --git a/counsel.el b/counsel.el
index bd1a4aa..b32d508 100644
--- a/counsel.el
+++ b/counsel.el
@@ -3911,7 +3911,7 @@ counsel-register-action
 Call a function on REGISTER.  The function is determined by
 matching the register's value description against a regexp in
 `counsel-register-actions'."
-  (setq register (string-to-char register))
+  (setq register (get-text-property 0 'register register))
   (let* ((desc (with-output-to-string
                  (register-val-describe (get-register register) nil)))
          (action (cl-assoc-if (lambda (re)
@@ -3926,11 +3926,14 @@ counsel-register
   "Interactively choose a register."
   (interactive)
   (ivy-read "Register: "
-            (mapcar (lambda (reg)
-                      (let* ((s (funcall register-preview-function reg))
-                             (i (string-match-p "\\(?:[ \t\n\r]+\\)\\'" s)))
-                        (if i (substring s 0 i) s)))
-                    register-alist)
+            (cl-mapcan
+             (lambda (reg)
+               (let ((s (funcall register-preview-function reg)))
+                 (setq s (substring s 0 (string-match-p "[ \t\n\r]+\\'" s)))
+                 (unless (string= s "")
+                   (put-text-property 0 1 'register (car reg) s)
+                   (list s))))
+             register-alist)
             :require-match t
             :sort t
             :history 'counsel-register-history
```